modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaBuilder.es.js

```diff
@@ -0,0 +1,315 @@
+	conjunctionName: PropTypes.string,
+	groupId: PropTypes.string,
+	items: PropTypes.array
+};
+
+const CRITERION_SHAPE = {
+	operatorName: PropTypes.string,
+	propertyName: PropTypes.string,
+	value: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
+};
+
+class CriteriaBuilder extends Component {
+	static propTypes = {
+		criteria: PropTypes.shape(
+			{
+				conjunctionName: PropTypes.string,
+				groupId: PropTypes.string,
+				items: PropTypes.arrayOf(
+					PropTypes.oneOfType(
+						[
+							PropTypes.shape(CRITERIA_GROUP_SHAPE),
+							PropTypes.shape(CRITERION_SHAPE)
+						]
+					)
+				)
+			}
+		),
+		modelLabel: PropTypes.string,
+		onChange: PropTypes.func,
+		supportedConjunctions: PropTypes.arrayOf(
+			PropTypes.shape(
+				{
+					label: PropTypes.string,
+					name: PropTypes.string.isRequired
+				}
+			)
+		),
+		supportedOperators: PropTypes.array,
+		supportedProperties: PropTypes.arrayOf(
+			PropTypes.shape(
+				{
+					entityUrl: PropTypes.string,
+					label: PropTypes.string,
+					name: PropTypes.string.isRequired,
+					options: PropTypes.array,
+					type: PropTypes.string.isRequired
+				}
+			)
+		).isRequired,
+		supportedPropertyTypes: PropTypes.object
+	};
+
+	static defaultProps = {
+		readOnly: false
+	};
+
+	state = {
+		editing: false
+	};
+
+	/**
+	 * Cleans the criteria by performing the following:
+	 * 1. Remove any groups with no items.
+	 * 2. Flatten groups that directly contain a single group.
+	 * 3. Flatten groups that contain a single criterion.
+	 * @param {array} criteriaItems A list of criterion and criteria groups
+	 * @param {boolean} root True if the criteriaItems are from the root group.
+	 * to clean.
+	 */
+	_cleanCriteriaMapItems(criteriaItems, root) {
+		return criteriaItems
+			.filter(
+				({items}) => {
+					return items ? items.length : true;
+				}
+			)
+			.map(
+				item => {
+					let cleanedItem = item;
+
+					if (item.items) {
+						if (item.items.length === 1) {
+							const soloItem = item.items[0];
+
+							if (soloItem.items) {
+								cleanedItem = {
+									conjunctionName: soloItem.conjunctionName,
+									groupId: soloItem.groupId,
+									items: this._cleanCriteriaMapItems(soloItem.items)
+								};
+							}
+							else {
+								cleanedItem = root ? item : soloItem;
+							}
+						}
+						else {
+							cleanedItem = {
+								...item,
+								...{
+									items: this._cleanCriteriaMapItems(item.items)
+								}
+							};
+						}
+					}
+
+					return cleanedItem;
+				}
+			);
+	}
+
+	/**
+	 * Switches the edit state between true and false.
+	 */
+	_handleToggleEdit = () => {
+		this.setState(
+			{
+				editing: !this.state.editing
+			}
+		);
+	};
+
+	/**
+	 * Cleans and updates the criteria with the newer criteria.
+	 * @param {Object} newCriteria The criteria with the most recent changes.
+	 */
+	_handleCriteriaChange = newCriteria => {
+		this.props.onChange(this._cleanCriteriaMapItems([newCriteria], true).pop());
```

Comment:
I'd avoid using `.pop()` unless you are intentionally wanting to mutate the array. It doesn't matter much here if you don't care about the array being mutated, but its probably a good habit to not use `.pop()` unless you are intentionally wanting to mutate.

I can be persuaded to keep the `pop()`, I am mostly just thinking of habits going forward with a functional react approach.

Example:
```js
const items = this._cleanCriteriaMapItems([newCriteria], true);

this.props.onChange(items[items.length - 1]);
```

Reply:
👍I agree with that. I'll make the changes.